### PR TITLE
Add and remove individual hands

### PR DESCRIPTION
**Pull request by stilnat**
## Summary

Following the inventory rework, this makes hands works properly when they are added or removed.
Currently, hands are controlled by a unique script, Hands.cs. This add another script, Hand.cs, representing a single hand.
Hands.cs becomes a controller for each hand on the player.

Server and client methods are also properly separated and varaibles synced, allowing only server to do things such as removing hands.

When a hand is removed, if it was the active hand, then another available hand become active, or none if there's none left.
It integrates nicely with interactions, as with no hands left it won't throw exceptions if the player tries to interact.

This is necessary to move on with basic health, as this will allow loosing hands properly.

## File changes

- LeftHand.prefab and rightHand.prefab now don't have the item component nor rigidbody.
- Adding two prefabs for left and right hands, basically a copy of the old ones.
- Added an assetdatabase for bodyparts (which generate 15+ extra files)

https://youtu.be/6u2Og_aA63w

## TODO

- [x] Add adding hands properly
- [x]  Add server command to test adding hands properly.

## Known issues

no known issues

## Fixes (optional)

no known fixes

### CODE REVIEW

**Comment by cosmiccoincidence** [comment]
Another 250 file PR..  oh boy.

**Comment by stilnat** [comment]
@cosmiccoincidence 

> (the huge amount of files changed is because it's based on https://github.com/RE-SS3D/SS3D/pull/1215), in reality, it's around 20 files changed.)

**Comment by cosmiccoincidence** [comment]
> @cosmiccoincidence
> 
> > (the huge amount of files changed is because it's based on #1215), in reality, it's around 20 files changed.)

Please base your new branches off your updated develop branch as opposed to your secondary branches which have already been used for other PRs, this causes the issues seen in this pr where there seem to be many more commits and files changed than were actually done in the PR and creates an inaccurate and confusing review. Its possible to merge other branches into your new branch still, so you can still move those commits into the branch you are working on.

**Comment by stilnat** [comment]
@cosmiccoincidence partly fixed for the number of files. I had to use my other branch as it was the prolongation of the inventory work. I should have waited for it to get merged before doing a pull request for this one sorry. I'll be more careful for the next one. (Plus I should have put WIP as I want to add a few stuff)

**Comment by cosmiccoincidence** [comment]
> @cosmiccoincidence partly fixed for the number of files. I had to use my other branch as it was the prolongation of the inventory work. I should have waited for it to get merged before doing a pull request for this one sorry. I'll be more careful for the next one. (Plus I should have put WIP as I want to add a few stuff)

Understood. At least you were able to solve the files part with a merge.

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
+﻿using SS3D.Core;
+using SS3D.Systems.Entities;
+using SS3D.Systems.Entities.Events;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+using SS3D.Systems.Permissions;
+using SS3D.Systems.PlayerControl;
+using System.Collections;
+using System.Collections.Generic;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+using UnityEngine;
+using FishNet;
+using UnityEditor;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+using SS3D.Data;
+using SS3D.Data.Enums;
+using FishNet.Object;
+using UnityEngine.InputSystem;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+namespace SS3D.Systems.IngameConsoleSystem.Commands
+{
+	/// <summary>
+	/// Command to add a hand to an entity.
+	/// This is mostly used for testing purpose, to check if hands can correctly be added to an entity and if they behave
+	/// as expected.
+	/// </summary>
+	public class AddHandCommand : Command
+	{
+		public override string LongDescription => "add hand to user (ckey), you can precise position (x,y,z) and rotation (x,y,z) argument ";
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
		public override string LongDescription => "add (ckey) [(position) (rotation)]\n Position and rotation are float arrays and written as x y z";
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+		public override string ShortDescription => "add hand to user";
+		public override ServerRoleTypes AccessLevel => ServerRoleTypes.Administrator;
+
+		public GameObject hand;
+
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		public override string Perform(string[] args)
+		{
+			CheckArgsResponse checkArgsResponse = CheckArgs(args);
+			if (checkArgsResponse.IsValid == false)
+				return checkArgsResponse.InvalidArgs;
+			string ckey = args[0];
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs:33]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+			Player Player = Subsystems.Get<PlayerSystem>().GetPlayer(ckey);
+			Entity entity = Subsystems.Get<EntitySystem>().GetSpawnedEntity(Player);
+
+			GameObject leftHandPrefab = Assets.Get<GameObject>((int)AssetDatabases.BodyParts, (int)BodyPartsIds.HumanHandLeft);
+			GameObject leftHandObject = GameObject.Instantiate(leftHandPrefab, entity.transform);
+			leftHandObject.transform.localPosition = position;
+			leftHandObject.transform.localEulerAngles = rotation;
+
+			Hand leftHand = leftHandObject.GetComponent<Hand>();
+			InstanceFinder.ServerManager.Spawn(leftHandObject, Player.Owner);
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
			Player player = Subsystems.Get<PlayerSystem>().GetPlayer(ckey);
			Entity entity = Subsystems.Get<EntitySystem>().GetSpawnedEntity(player);

			GameObject leftHandPrefab = Assets.Get<GameObject>((int)AssetDatabases.BodyParts, (int)BodyPartsIds.HumanHandLeft);
			GameObject leftHandObject = Object.Instantiate(leftHandPrefab, entity.transform);
			leftHandObject.transform.localPosition = position;
			leftHandObject.transform.localEulerAngles = rotation;

			Hand leftHand = leftHandObject.GetComponent<Hand>();
			InstanceFinder.ServerManager.Spawn(leftHandObject, player.Owner);
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/Command.cs [hunk 9]
```diff
@@ -1,4 +1,5 @@
-﻿using SS3D.Systems.Permissions;
+﻿using FishNet.Object;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/Command.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+		}
+
+		[Server]
+		private void CmdPerform(string[] args, NetworkConnection conn = null)
+		{
+			string ckey = args[0];
+			if(!Subsystems.Get<PermissionSystem>().TryGetUserRole(args[0], out ServerRoleTypes userPermission))
+			{
+				return;
+			}
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]: Should be in CheckArgs()

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+			if(userPermission < AccessLevel)
+			{
+				return;
+			}
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs]:
```suggestion
```
Will be in CommandsController

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+
+		/// <summary>
+		/// the hands script controlling this hand.
+		/// </summary>
+		public Hands handsController;
+
+		public Vector3 InteractionOrigin => interactionOrigin.position;
+
+		public delegate void HandHandler(Hand hand);
+		public event HandHandler OnHandDisabled;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs]: Fix naming of the fields

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+		public bool IsEmpty()
+		{
+			return Container.Container.Empty;
+		}
+
+		/// <summary>
+		/// Get the interaction source from stuff in hand if there's any.
+		/// </summary>
+		/// <returns></returns>
+		public IInteractionSource GetActiveTool()
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:69]: Can you explain this? What is interaction source?

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+			return interactionSource;
+		}
+
+		public RangeLimit GetInteractionRange()
+		{
+			return range;
+		}
+
+		[Server]
+		public void Pickup(Item item)
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:91]: Probably should be [ServerRpc]

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs [hunk 15]
```diff
@@ -0,0 +1,145 @@
+﻿using SS3D.Systems.Inventory.Containers;
+using System.Collections;
+using System.Collections.Generic;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+using UnityEngine;
+using SS3D.Systems.Inventory.Items;
+using System.Linq;
+using SS3D.Interactions.Interfaces;
+using SS3D.Interactions;
+using FishNet.Object;
+using static SS3D.Systems.Inventory.Containers.AttachedContainer;
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs]:
```suggestion
```

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/InventoryView.cs [hunk 17]
```diff
@@ -344,26 +343,10 @@ private SingleItemContainerSlot AddHorizontalLayoutSlot(GameObject prefab, Conta
...
-            }
-
-            if(index+childIndex >= HorizontalLayout.transform.childCount)
-            {
-                Punpun.Warning(this, "index out of bound, check that the number of hand slots is greater than index.");
-                return null;
-            }
-
-            return HorizontalLayout.transform.GetChild(index+childIndex);
+			var slot = Slots.Find(x => x.Container == hand.Container);
```
Comment (iamteapot422) [inline Assets/Scripts/SS3D/Systems/Inventory/Containers/InventoryView.cs]:
```suggestion
			SingleItemContainerSlot slot = Slots.Find(x => x.Container == hand.Container);
```

**Review by iamteapot422 — CHANGES_REQUESTED** [review]

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs (hunk 13 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:69]: updated IInteractionSource doc, lmk if this helps enough

File: Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs (hunk 7 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/IngameConsoleSystem/Commands/AddHandCommand.cs:33]: I need ckey it to find the right player

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs (hunk 12 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs]: not sure what you expect, I tried a renaming, you'll tell me if that works

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs (hunk 14 above)
Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:91]: This should be called only on server by server I think. Its working with [server] even on client. The interaction system lacks clarity regarding client and server code.

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs (hunk 12 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs]: Change range to Range, pickupIcon to PickupIcon, interactionOrigin to InteractionOrigin, handsController to HandsController

File: Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs (hunk 13 above)
Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:69]: If a screwdriver can be a source for opening panels, then why interactionSource.Source = this; sets a hand as the source?

Comment (stilnat) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:69]: @iamteapot422 the method check if an item held in hand should be used as the interaction source, such as a screwdriver or a hammer or whatever... If nothing is found, the hand itself is the interaction source, hands have their own set of available interactions (hit, pickup... more coming)

Comment (iamteapot422) [reply Assets/Scripts/SS3D/Systems/Inventory/Containers/Hand.cs:69]: @stilnat shouldn't the check be "if (interactionSource == null)" then?

**Review by iamteapot422 — APPROVED** [review]

**Pull request merged**